spec/unit/QueuePool.spec.js

```diff
@@ -0,0 +1,138 @@
+        });
+    });
+}
+
+describe("QueuePool", function() {
+    const size = 3;
+    let pool;
+    let procFn;
+    let itemToDeferMap;
+
+    let resolveItem = function(id, resolveWith) {
+        if (!itemToDeferMap[id]) {
+            return;
+        }
+        itemToDeferMap[id].resolve(resolveWith);
+        delete itemToDeferMap[id];
+    }
+
+    beforeEach(function() {
+        procFn = jasmine.createSpy("procFn");
+        pool = new QueuePool(size, procFn);
+        itemToDeferMap = {
+            // $item: Deferred
+        };
+        procFn.andCallFake((item) => {
+            itemToDeferMap[item] = new promiseutil.defer();
+            return itemToDeferMap[item].promise;
+        })
+    });
+
+    it("should let multiple items be processed at once",
+    test.coroutine(function*() {
+        pool.enqueue("a", "a");
+        pool.enqueue("b", "b");
+        // procFn is called on the next tick so check they've been called after
+        yield nextTick();
+        expect(Object.keys(itemToDeferMap).length).toBe(2);
+    }));
+
+    it("should resolve enqueued items when they resolve",
+    test.coroutine(function*() {
+        pool.enqueue("a", "a");
+        let promise = pool.enqueue("b", "b");
+        yield nextTick();
+        resolveItem("b", "stuff");
+        let res = yield promise;
+        expect(res).toEqual("stuff");
+    }));
+
+    it("should not let more items than the pool size be processed at once",
+    test.coroutine(function*() {
+        pool.enqueue("a", "a");
+        pool.enqueue("b", "b");
+        pool.enqueue("c", "c");
+        pool.enqueue("d", "d");
+        yield nextTick();
+        expect(Object.keys(itemToDeferMap).sort()).toEqual(["a", "b", "c"]);
+        resolveItem("b");
+        yield nextTick();
+        expect(Object.keys(itemToDeferMap).sort()).toEqual(["a", "c", "d"]);
+    }));
+
+    it("should wait until a queue is free", test.coroutine(function*() {
+        pool.enqueue("a", "a");
+        pool.enqueue("b", "b");
+        pool.enqueue("c", "c");
+        yield nextTick();
+        expect(Object.keys(itemToDeferMap).sort()).toEqual(["a", "b", "c"]);
+        yield nextTick();
+        yield nextTick();
+        pool.enqueue("d", "d");
+        // wait a while
+        yield nextTick();
+        yield nextTick();
+        yield nextTick();
+        yield nextTick();
+        expect(Object.keys(itemToDeferMap).sort()).toEqual(["a", "b", "c"]);
+        resolveItem("c");
+        yield nextTick();
+        expect(Object.keys(itemToDeferMap).sort()).toEqual(["a", "b", "d"]);
+    }));
+
+    it("should process overflows FIFO", test.coroutine(function*() {
+        pool.enqueue("a", "a");
+        pool.enqueue("b", "b");
+        pool.enqueue("c", "c");
+        pool.enqueue("d", "d");
+        pool.enqueue("e", "e");
+        yield nextTick();
+        expect(Object.keys(itemToDeferMap).sort()).toEqual(["a", "b", "c"]);
+        pool.enqueue("f", "f");
+        resolveItem("b");
```

Comment:
Nope. Just tried re-ordering (resolve then enqueue before the next tick) and it works fine. The overflow itself is a `Queue` so won't be prone to race conditions like this.